File: python/src/fem.py
import numpy as np
import concurrent.futures
import sys

class Fem():
    def __init__(self,dof,nodes,elements,materials):
        self.nnode = len(nodes)
        self.nelem = len(elements)
        self.dof = dof

        self.nodes = nodes
        self.elements = elements
        self.materials = materials

        self.free_nodes = []
        self.fixed_nodes = []

        self.input_elements = []
        self.connected_elements = []
        self.spring_elements = []

        self.e_elements = []
        self.ep_elements = []
        self.ep_eff_elements = []

    # ======================================================================= #
    def set_init(self):
        self._set_mesh()
        self._set_initial_condition()
        self._set_initial_matrix()

    # ---------------------------------------
    def _set_mesh(self):
        for element in self.elements:
            nodes = []
            for inode in element.inode:
                if self.nodes[inode].id == inode:
                    nodes += [self.nodes[inode]]
                else:
                    for n in self.nodes:
                        if n.id == inode:
                            nodes += [n]
                            break
            element.set_nodes(nodes)

            if element.material_id < 0:
                element.set_material(self.dof,None)
            else:
                if self.materials[element.material_id].id == element.material_id:
                    material = self.materials[element.material_id]
                else:
                    for m in self.materials:
                        if m.id == element.material_id:
                            material = m
                            break

                element.set_material(self.dof,material)

                if "ep_eff_" in material.style:
                    self.ep_eff_elements += [element]
                elif "ep_" in material.style:
                    self.ep_elements += [element]
                else:
                    self.e_elements += [element]

            if "input" in element.style:
                self.input_elements += [element]
            if "connect" in element.style:
                self.connected_elements += [element]
            if "spring" in element.style:
                self.spring_elements += [element]

    # ---------------------------------------
    def _set_initial_condition(self):
        for node in self.nodes:
            node.set_initial_condition()

            if np.count_nonzero(node.freedom) == self.dof:
                self.free_nodes += [node]
            else:
                self.fixed_nodes += [node]

        for element in self.elements:
            element.set_pointer_list()

    # ---------------------------------------
    def _set_initial_matrix(self):
        for element in self.elements:
            element.set_xn()
            element.mk_local_matrix_init(self.dof)
            element.mk_local_matrix()
            element.mk_local_vector()

            id = 0
            for node in element.nodes:
                for i in range(self.dof):
                    node.mass[i] += element.M_diag[id]
                    node.c[i] += element.C_diag[id]
                    node.k[i] += element.K_diag[id]
                    node.static_force[i] += element.force[id]
                    id += 1

    # ======================================================================= #
    def set_output(self,outputs):
        output_node_list,output_element_list = outputs

        self.output_nnode = len(output_node_list)
        self.output_nodes = [None] * self.output_nnode
        for id,inode in enumerate(output_node_list):
            self.output_nodes[id] = self.nodes[inode]

        self.output_nelem = len(output_element_list)
        self.output_elements = [None] * self.output_nelem
        self.output_e_elements,self.output_ep_elements = [],[]
        for id,ielem in enumerate(output_element_list):
            self.output_elements[id] = self.elements[ielem]
            if "ep_" in self.elements[ielem].material.style:
                self.output_ep_elements += [self.elements[ielem]]
            else:
                self.output_e_elements += [self.elements[ielem]]

    # ======================================================================= #
    def set_rayleigh_damping(self,f0,f1,h):
        omega0,omega1 = 2*np.pi*f0, 2*np.pi*f1
        alpha = 2*h*omega0*omega1 / (omega0+omega1)
        beta = 2*h / (omega0+omega1)
        print(alpha,beta)

        for node in self.nodes:
            node.c    = np.zeros(self.dof,dtype=np.float64)

        for element in self.elements:
            element.mk_local_damping_matrix(alpha,beta)

            id = 0
            for node in element.nodes:
                for i in range(self.dof):
                    node.c[i] += element.C_diag[id]
                    id += 1


    # ======================================================================= #
    def self_gravity(self):
        ### Initial condition ###
        H = 0.0
        for node in self.nodes:
            if node.xyz[1] > H:
                H = node.xyz[1]

        g,vp = 9.8,1500.0
        for node in self.node_set:
            node.u[0] = 0.0
            node.u[1] = g/(2*vp**2) * (H**2 - node.xyz[1]**2)
            node.um = np.copy(node.u)

        self._self_gravity_cg(full=False)
        self._self_gravity_cg(full=True)

        for node in self.node_set:
            node.u0 = np.copy(node.u)
            node.um = np.copy(node.u)

    # ---------------------------------------
    def _self_gravity_cg(self,full=True,print_flag=True):
        if full:    # Calculate both components
            id = 0
        else:       # Calculate only vertical component
            id = 1

        ### CG Method ###
        for node in self.nodes:
            node.force = np.zeros(node.dof,dtype=np.float64)
        for element in self.elements:
            element.mk_ku()
        for node in self.nodes:
            for i in range(id,node.dof):
                if node.freedom[i] == 0:
                    node._ur[i] = 0.0
                else:
                    node._ur[i] = node.static_force[i] - node.force[i]
        for element in self.connected_elements:
            u = np.zeros(element.nodes[0].dof,dtype=np.float64)
            for node in element.nodes:
                u += node._ur
            for node in element.nodes:
                node._ur = u/element.nnode
        for element in self.input_elements:
            for node in element.nodes:
                node._ur = np.zeros(element.nodes[0].dof,dtype=np.float64)
        for node in self.nodes:
            node._up = np.copy(node._ur)
        for element in self.elements:
            element._up = ()
            for node in element.nodes:
                element._up += (node._up.view(),)

        for it in range(10*self.nnode):
            ## y = Ap
            for node in self.nodes:
                node.force = np.zeros(node.dof,dtype=np.float64)
            for element in self.elements:
                element.mk_ku_u(element._up)
            for node in self.nodes:
                node._uy = node.force

            ## correction boundary condition
            for node in self.nodes:
                for i in range(id,node.dof):
                    if node.freedom[i] == 0:
                        node._uy[i] = 0.0
            for element in self.connected_elements:
                u = np.zeros(element.nodes[0].dof,dtype=np.float64)
                for node in element.nodes:
                    u += node._uy
                for node in element.nodes:
                    node._uy = u/element.nnode
            for element in self.input_elements:
                for node in element.nodes:
                    node._uy = np.zeros(element.nodes[0].dof,dtype=np.float64)


            ## alpha = rr/py
            rr,py = 0.0,0.0
            for node in self.nodes:
                rr += node._ur @ node._ur
                py += node._up @ node._uy
            alpha = rr/py

            ## x = x + alpha*p
            rr1 = 0.0
            for node in self.nodes:
                for i in range(id,node.dof):
                    if node.freedom[i] == 0:
                        pass
                    else:
                        node.u[i] += alpha*node._up[i]
                        node._ur[i] -= alpha*node._uy[i]
                rr1 += node._ur @ node._ur

            if rr1 < 1.e-10:
                if print_flag:
                    print(" (self gravity process .. )",it,self.nodes[0].u[1],rr1)
                break

            ## p = r + beta*p
            beta = rr1/rr
            for node in self.nodes:
                for i in range(id,node.dof):
                    if node.freedom[i] == 0:
                        pass
                    else:
                        node._up[i] = node._ur[i] + beta*node._up[i]

            if it%100 == 0 and print_flag:
                print(" (self gravity process .. )",it,self.nodes[0].u[1],rr1)

    # ======================================================================= #
    def set_ep_initial_state(self):
        self._self_gravity_cg(full=False,print_flag=False)

        for element in self.ep_elements:
            element.calc_stress()
            # element.ep.initial_state_isotropic(element.stress)
            element.ep.initial_state(element.stress)
            element.material.rmu,element.material.rlambda = element.ep.elastic_modulus()
            element.clear_strain()

        for element in self.ep_eff_elements:
            element.calc_eff_stress()
            # element.eff_stress = [-30.e3,-30.e3,0]    #等方圧密
            # element.ep.initial_state_isotropic(element.eff_stress)
            element.ep.initial_state(element.eff_stress)
            element.material.rmu,element.material.rlambda = element.ep.elastic_modulus()
            element.clear_strain()
            element.stress = np.copy(element.eff_stress)

        self._set_ep_initial_state_node_clear()

        for node in self.nodes:
            node.u[:] = np.zeros(self.dof,dtype=np.float64)

        self._set_initial_matrix()

        for element in self.ep_elements:
            element.ep_init_calc_stress_all()
            print("e is",element.ep.e)
        for element in self.ep_eff_elements:
            element.ep_eff_init_calc_stress_all()

    # ---------------------------------------
    def _set_ep_initial_state_node_clear(self):
        for node in self.nodes:
            node.mass = np.zeros(self.dof,dtype=np.float64)
            node.c    = np.zeros(self.dof,dtype=np.float64)
            node.k      = np.zeros(self.dof,dtype=np.float64)
            node.force = np.zeros(self.dof,dtype=np.float64)
            node.static_force = np.zeros(self.dof,dtype=np.float64)
            node.dynamic_force = np.zeros(self.dof,dtype=np.float64)

    # ======================================================================= #
    def update_init(self,dt):
        for node in self.nodes:
            node.inv_mc = 1.0 / (node.mass[:] + 0.5*dt*node.c[:])
            node.mass_inv_mc = node.mass[:]*node.inv_mc[:]
            node.c_inv_mc = node.c[:]*node.inv_mc[:]*0.5*dt
            node.dtdt_inv_mc = dt*dt*node.inv_mc[:]

        self.dt = dt
        self.inv_dt2 = 1./(2.*dt)
        self.inv_dtdt = 1./(dt*dt)

    # ======================================================================= #
    def update_matrix(self):
        for node in self.nodes:
            self._update_matrix_node_init(node)
        for element in self.elements:
            self._update_matrix_set_elements(element)

    # ---------------------------------------
    def _update_matrix_node_init(self,node):
        node.mass = np.zeros(self.dof,dtype=np.float64)
        node.c    = np.zeros(self.dof,dtype=np.float64)
        node.dynamic_force = np.zeros(self.dof,dtype=np.float64)

    def _update_matrix_set_elements(self,element):
        element.set_xn()
        element.mk_local_update()

        id = 0
        for node in element.nodes:
            for i in range(self.dof):
                node.mass[i] += element.M_diag[id]
                node.c[i] += element.C_diag[id]
                node.dynamic_force[i] += element.force[id]
                id += 1

    def _update_matrix_set_nodes(self,node):
        node.inv_mc = 1.0 / (node.mass[:] + 0.5*self.dt*node.c[:])
        node.mass_inv_mc = node.mass[:]*node.inv_mc[:]
        node.c_inv_mc = node.c[:]*node.inv_mc[:]*0.5*self.dt
        node.dtdt_inv_mc = self.dt*self.dt*node.inv_mc[:]

    # ======================================================================= #
    def update_time(self,acc0,vel0=None,input_wave=False,self_gravity=False,FD=False):
        if FD:
            self.update_matrix()
        else:
            if self_gravity:
                for node in self.nodes:
                    node.dynamic_force = np.copy(node.static_force)
            else:
                for node in self.nodes:
                    node.dynamic_force = np.zeros(self.dof,dtype=np.float64)

        for node in self.nodes:
            self._update_time_node_init(node)

        if input_wave:
            for element in self.input_elements:
                self._update_time_input_wave(element,vel0)
        else:
            for element in self.elements:
                self._update_bodyforce(element,acc0)

        if FD:
            for element in self.e_elements:
                element.mk_B_stress()
                element.mk_cv()
            for element in self.ep_elements:
                element.mk_ep_FD_B_stress()
            for element in self.ep_eff_elements:
                element.mk_ep_FD_eff_B_stress()
        else:
            for element in self.e_elements:
                element.mk_ku_cv()
            for element in self.ep_elements:
                element.mk_ep_B_stress()
            for element in self.ep_eff_elements:
                element.mk_ep_eff_B_stress()

        self._update_time_set_nodes_all()

        for element in self.connected_elements:
            self._update_time_set_connected_elements_(element)
        for element in self.spring_elements:
            self._update_time_set_spring_elements_(element)

        if FD:
            for element in self.output_e_elements:
                element.calc_FD_stress()
        else:
            for element in self.output_e_elements:
                element.calc_stress()

    # ======================================================================= #
    def update_time_input(self,vel0):
        for node in self.node_set:
            node.dynamic_force = np.zeros(self.dof,dtype=np.float64)
            node.force = np.zeros(self.dof,dtype=np.float64)

        for element in self.input_element_set:
            self._update_time_input_wave(element,vel0)

        self._update_mk_internal_force()
        self._update_time_set_nodes_all()

        for element in self.connected_element_set:
            self._update_time_set_connected_elements_(element)
        for element in self.spring_elements:
            self._update_time_set_spring_elements_(element)

        for element in self.output_e_element_set:
            element.calc_stress()
        for element in self.output_ep_element_set:
            element.calc_ep_stress()

    # ---------------------------------------
    def _update_time_node_init(self,node):
        node.force = -node.dynamic_force.copy()

    def _update_time_input_wave(self,element,vel0):
        cv = element.C @ np.tile(vel0,element.nnode)
        for i in range(element.nnode):
            i0 = self.dof*i
            element.nodes[i].force[:] -= 2*cv[i0:i0+self.dof]

    def _update_bodyforce(self,element,acc0):
        element.mk_bodyforce(acc0)
        for i in range(element.nnode):
            i0 = self.dof*i
            element.nodes[i].force[:] -= element.force[i0:i0+self.dof]

    def _update_mk_internal_force(self):
        for element in self.elements:
            element.mk_ku_cv()

    def _update_time_set_nodes_all(self):
        for node in self.free_nodes:
            self._update_time_set_free_nodes(node)
        for node in self.fixed_nodes:
            self._update_time_set_fixed_nodes(node)

    def _update_time_set_free_nodes(self,node):
        u = np.copy(node.u)
        node.u[:] = node.mass_inv_mc*(2.*u-node.um) + node.c_inv_mc*node.um - node.dtdt_inv_mc*node.force
        node.v[:] = (node.u - node.um) * self.inv_dt2
        node.a[:] = (node.u - 2.*u + node.um) * self.inv_dtdt
        node.um = u

    def _update_time_set_fixed_nodes(self,node):
        u = np.copy(node.u)
        for i in range(self.dof):
            if node.freedom[i] == 0:
                node.u[i]  = 0.0
            else:
                node.u[i] = node.mass_inv_mc[i]*(2.*u[i]-node.um[i]) + node.c_inv_mc[i]*node.um[i] - node.dtdt_inv_mc[i]*node.force[i]
        node.v[:] = (node.u - node.um) * self.inv_dt2
        node.a[:] = (node.u - 2.*u + node.um) * self.inv_dtdt
        node.um = u

    def _update_time_set_connected_elements_(self,element):
        u = np.zeros_like(element.nodes[0].u)
        a = np.zeros_like(element.nodes[0].a)
        for node in element.nodes:
            u[:] += node.u[:]
            a[:] += node.a[:]
        for node in element.nodes:
            node.u[:] = u[:]/element.nnode
            node.a[:] = a[:]/element.nnode

    def _update_time_set_spring_elements_(self,element):
        u = element.R @ np.hstack(element.u)
        element.f[0] = element.material.kv*(u[0]-u[2])
        element.f[1] = element.material.kh*(u[1]-u[3])

    # ---------------------------------------
    def _multi_update_mk_internal_force(self):
        def mk_ku_cv(element):
            element.mk_ku_cv()
        with concurrent.futures.ProcessPoolExecutor() as executor:
            executor.map(mk_ku_cv,self.elements)

    def _multi_update_time_set_nodes_all(self):
        with concurrent.futures.ProcessPoolExecutor() as executor:
            executor.map(self._update_time_set_free_nodes,self.free_nodes)
        for node in self.fixed_node_set:
            self._update_time_set_fixed_nodes(node)

    # ======================================================================= #
    def print_all(self):
        for node in self.nodes:
            node.print()
        for element in self.elements:
            element.print()

File: cpp/src/fem.cpp
#include "all.h"
#include <Eigen/Core>
#include "node.h"
#include "material.h"
#include "element_style.h"
#include "element.h"
#include "fem.h"

using EV = Eigen::VectorXd ;

// ------------------------------------------------------------------- //
// ------------------------------------------------------------------- //
Fem::Fem (size_t dof, std::vector<Node> nodes,
                std::vector<Element> elements,
                std::vector<Material> materials)
  {
    this->nnode = nodes.size();
    this->nelem = elements.size();
    this->dof = dof;

    this->nodes = nodes;
    this->elements = elements;
    this->materials = materials;
  }

// ------------------------------------------------------------------- //
// ------------------------------------------------------------------- //
void Fem::set_init() {
    this->_set_mesh();
    this->_set_initial_condition();
    this->_set_initial_matrix();
  }

// ----------------------------- //
void Fem::_set_mesh() {
    for (auto& element : this->elements) {

      std::vector<Node*> nodes_p(element.nnode);
      for (size_t i = 0 ; i < element.nnode ; i++ ){
        size_t inode = element.inode[i];

        if (this->nodes[inode].id == inode) {
          nodes_p[i] = &this->nodes[inode];
        } else {
          for (auto& node : this->nodes) {
            if (node.id == inode) {
              nodes_p[i] = &node;
              break;
            }
          }
        }

      }

      element.set_nodes(nodes_p);

      Material* material_p = nullptr;
      if (this->materials[element.material_id].id == element.material_id) {
        material_p = &this->materials[element.material_id];
      } else {
        for (auto& material : this->materials) {
          if (material.id == element.material_id) {
            material_p = &material;
            break;
          }
        }
      }

      element.set_material(material_p);

      if (element.style.find("input") != std::string::npos) {
        this->input_elements_p.push_back(&element);
      }
      if (element.style.find("connect") != std::string::npos) {
        this->connected_elements_p.push_back(&element);
      }
    }
  }

// ----------------------------- //
void Fem::_set_initial_condition() {
    for (auto& node : this->nodes) {
      node.set_initial_condition();

      size_t total_dof = std::accumulate(node.freedom.begin(),node.freedom.end(),0);
      if (total_dof == dof) {
        this->free_nodes_p.push_back(&node);
      } else {
        this->fixed_nodes_p.push_back(&node);
      }
    }
  }

// ----------------------------- //
void Fem::_set_initial_matrix(){
    for (auto& element : this->elements) {
      element.set_xn();
      element.mk_local_matrix_init(this->dof);
      element.mk_local_matrix();
      element.mk_local_vector();

      size_t id = 0;
      for (size_t inode = 0 ; inode < element.nnode ; inode++) {
        for (size_t i = 0 ; i < this->dof ; i++) {
          element.nodes_p[inode]->mass[i] += element.M_diag[id];
          element.nodes_p[inode]->c[i] += element.C_diag[id];
          element.nodes_p[inode]->k[i] += element.K_diag[id];
          element.nodes_p[inode]->static_force[i] += element.force[id];
          id++;
        }
      }

    }
  }

// ------------------------------------------------------------------- //
// ------------------------------------------------------------------- //
void Fem::set_output(std::tuple<std::vector<size_t>, std::vector<size_t>> outputs) {
    auto [output_node_list, output_element_list] = outputs;

    this->output_nnode = output_node_list.size();
    for (size_t inode = 0 ; inode < this->output_nnode ; inode++) {
      size_t id = output_node_list[inode];
      this->output_nodes_p.push_back(&this->nodes[id]);
    }

    this->output_nelem = output_element_list.size();
    for (size_t ielem = 0 ; ielem < this->output_nelem ; ielem++) {
      size_t id = output_element_list[ielem];
      this->output_elements_p.push_back(&this->elements[id]);
    }
  }

// ------------------------------------------------------------------- //
// ------------------------------------------------------------------- //
void Fem::self_gravity() {
    // Initial condition //
    double H = 0.0;
    for (auto& node : this->nodes) {
      if (node.xyz[1] > H) {
        H = node.xyz[1];
      }
    }

    double g = 9.8;
    double vp = 1500.0;
    for (auto& node : this->nodes) {
      node.u[0] = 0.0;
      node.u[1] = g/(2*vp*vp) * (H*H - node.xyz[1]*node.xyz[1]);
      node.um = node.u;
    }

    this->_self_gravity_cg(false);
    this->_self_gravity_cg(true);

    for (auto& node : this->nodes) {
      node.u0 = node.u;
      node.um = node.u;
    }

  }

// ------------------------------------------------------------------- //
void Fem::_self_gravity_cg(const bool full) {
    size_t id;
    if (full) {
      id = 0;
    } else {
      id = 1;
    }

    // CG Method //
    // --- set initial variables --- //
    for (auto& node : this->nodes) {
      node.force = EV::Zero(node.dof);
    }
    for (auto& element : this->elements) {
      element.mk_ku();
    }
    for (auto& node : this->nodes) {
      for (size_t i = id ; i < node.dof ; i++) {
        if (node.freedom[i] == 0) {
          node._ur[i] = 0.0;
        } else {
          node._ur[i] = node.static_force[i] - node.force[i];
        }
      }
    }
    for (auto& element_p : this->connected_elements_p) {
      EV u = EV::Zero(element_p->dof);
      for (size_t inode = 0 ; inode < element_p->nnode ; inode++) {
        u += element_p->nodes_p[inode]->_ur;
      }
      for (size_t inode = 0 ; inode < element_p->nnode ; inode++) {
        element_p->nodes_p[inode]->_ur = u/element_p->nnode;
      }
    }
    for (auto& element_p : this->input_elements_p) {
      for (size_t inode = 0 ; inode < element_p->nnode ; inode++) {
        element_p->nodes_p[inode]->_ur = EV::Zero(element_p->dof);
      }
    }
    for (auto& node : this->nodes) {
      node._up = node._ur;
    }


    // --- CG iterations --- //
    for (size_t it=0 ; it < 10*this->nnode ; it++) {
      double rr, rr1, py, alpha, beta;

      // y = Ap
      for (auto& node : this->nodes) {
        node.force = EV::Zero(node.dof);
      }
      for (auto& element : this->elements) {
        element.mk_ku_up();
      }
      for (auto& node : this->nodes) {
        node._uy = node.force;
      }

      // correction boundary condition
      for (auto& node : this->nodes) {
        for (size_t i = id ; i < node.dof ; i++) {
          if (node.freedom[i] == 0) {
            node._uy[i] = 0.0;
          }
        }
      }
      for (auto& element_p : this->connected_elements_p) {
        EV u = EV::Zero(element_p->dof);
        for (size_t inode = 0 ; inode < element_p->nnode ; inode++) {
          u += element_p->nodes_p[inode]->_uy;
        }
        for (size_t inode = 0 ; inode < element_p->nnode ; inode++) {
          element_p->nodes_p[inode]->_uy = u/element_p->nnode;
        }
      }
      for (auto& element_p : this->input_elements_p) {
        for (size_t inode = 0 ; inode < element_p->nnode ; inode++) {
          element_p->nodes_p[inode]->_uy = EV::Zero(element_p->dof);
        }
      }

      // alpha = rr/py
      rr = 0.0; py = 0.0;
      for (auto& node : this->nodes) {
        rr += node._ur.dot(node._ur);
        py += node._up.dot(node._uy);
      }
      alpha = rr/py;

      // x = x + alpha*p
      rr1 = 0.0;
      for (auto& node : this->nodes) {
        for (size_t i = id ; i < node.dof ; i++) {
          if (node.freedom[i] != 0) {
            node.u[i] += alpha * node._up[i];
            node._ur[i] -= alpha * node._uy[i];
          }
        }
        rr1 += node._ur.dot(node._ur);
      }

      if (rr1 < 1.e-10) {
        std::cout << " (self gravity process .. ) " << it << " " ;
        std::cout << this->nodes[0].u[1] << " ";
        std::cout << rr1 << "\n";
        break;
      }

      // p = r + beta*p
      beta = rr1/rr;
      for (auto& node : this->nodes) {
        for (size_t i = id ; i < node.dof ; i++) {
          if (node.freedom[i] != 0) {
            node._up[i] = node._ur[i] + beta * node._up[i];
          }
        }
      }

      if (it%100 == 0){
        std::cout << " (self gravity process .. ) " << it << " " ;
        std::cout << this->nodes[0].u[1] << " ";
        std::cout << rr1 << "\n";
      }
    }

  }

// ------------------------------------------------------------------- //
// ------------------------------------------------------------------- //
void Fem::update_init(const double dt) {
    for (auto& node : this->nodes) {
      for (size_t i = 0 ; i < node.dof ; i++) {
        node.inv_mc[i] = 1.0 / (node.mass[i] + 0.5*dt*node.c[i]);
        node.mass_inv_mc[i] = node.mass[i] * node.inv_mc[i];
        node.c_inv_mc[i] = node.c[i] * node.inv_mc[i] * 0.5*dt;
        node.dtdt_inv_mc[i] = dt*dt*node.inv_mc[i];
      }
    }
    this->dt = dt;
    this->inv_dt2 = 1.0/(2.0*dt);
  }

// ------------------------------------------------------------------- //
// ------------------------------------------------------------------- //
void Fem::update_time(const EV acc0, const EV vel0, const bool input_wave, const bool FD) {
    if(FD) {
      if(input_wave) {
        this->update_time_input_FD(vel0);
      } else {
        this->update_time_FD(acc0);
      }
    } else {
      if(input_wave) {
        this->update_time_input_MD(vel0);
      } else {
        this->update_time_MD(acc0);
      }
    }
  }

// ------------------------------------------------------------------- //
void Fem::update_time_FD(const EV acc0) {

    this->_update_matrix();

    for (auto& node : this->nodes) {
      node.force = -node.dynamic_force;
    }

    for (auto& element : this->elements) {
      element.update_bodyforce(acc0);
    }

    for (auto& element : this->elements) {
      element.mk_B_stress();
      element.mk_cv();
    }

    this->_update_time_set_free_nodes();
    this->_update_time_set_fixed_nodes();
    this->_update_time_set_connected_elements();

    for (auto& element_p : this->output_elements_p) {
      element_p->calc_stress();
    }
  }

// ------------------------------------------------------------------- //
void Fem::update_time_input_FD(const EV vel0) {

    this->_update_matrix();

    for (auto& node : this->nodes) {
      node.force = -node.dynamic_force;
    }

    for (auto& element_p : this->input_elements_p) {
      element_p->update_inputwave(vel0);
    }

    for (auto& element : this->elements) {
      element.mk_B_stress();
      element.mk_cv();
    }

    this->_update_time_set_free_nodes();
    this->_update_time_set_fixed_nodes();
    this->_update_time_set_connected_elements();

    for (auto& element_p : this->output_elements_p) {
      element_p->calc_stress();
    }
  }


// ------------------------------------------------------------------- //
void Fem::update_time_MD(const EV acc0) {
    for (auto& node : this->nodes) {
      node.force = -node.dynamic_force;
    }

    for (auto& element : this->elements) {
      element.update_bodyforce(acc0);
    }

    for (auto& element : this->elements) {
      element.mk_ku_cv();
    }

    this->_update_time_set_free_nodes();
    this->_update_time_set_fixed_nodes();
    this->_update_time_set_connected_elements();

    for (auto& element_p : this->output_elements_p) {
      element_p->calc_stress();
    }
  }

// ------------------------------------------------------------------- //
void Fem::update_time_input_MD(const EV vel0) {
    for (auto& node : this->nodes) {
      node.force = -node.dynamic_force;
    }

    for (auto& element_p : this->input_elements_p) {
      element_p->update_inputwave(vel0);
    }

    for (auto& element : this->elements) {
      element.mk_ku_cv();
    }

    this->_update_time_set_free_nodes();
    this->_update_time_set_fixed_nodes();
    this->_update_time_set_connected_elements();

    for (auto& element_p : this->output_elements_p) {
      element_p->calc_stress();
    }
  }

// ------------------------------------------------------------------- //
// ------------------------------------------------------------------- //
void Fem::_update_time_set_free_nodes() {
    for (auto& node_p : this->free_nodes_p) {
      EV u = node_p->u;
      for (size_t i = 0 ; i < node_p->dof ; i++) {
        node_p->u[i] = node_p->mass_inv_mc[i]*(2.0*u[i] - node_p->um[i])
                + node_p->c_inv_mc[i]*node_p->um[i] - node_p->dtdt_inv_mc[i]*node_p->force[i];
      }
      node_p->v = (node_p->u - node_p->um) * this->inv_dt2;
      node_p->um = u;
    }
  }

void Fem::_update_time_set_fixed_nodes() {
    for (auto& node_p : this->fixed_nodes_p) {
      EV u = node_p->u;
      for (size_t i = 0 ; i < node_p->dof ; i++) {
        if (node_p->freedom[i] == 0) {
          node_p->u[i] = 0.0;
        } else {
          node_p->u[i] = node_p->mass_inv_mc[i]*(2.0*u[i] - node_p->um[i])
                  + node_p->c_inv_mc[i]*node_p->um[i] - node_p->dtdt_inv_mc[i]*node_p->force[i];
        }
      }
      node_p->v = (node_p->u - node_p->um) * this->inv_dt2;
      node_p->um = u;
    }
  }

void Fem::_update_time_set_connected_elements() {
    for (auto& element_p : this->connected_elements_p) {
      EV u = EV::Zero(element_p->dof);
      for (size_t inode = 0 ; inode < element_p->nnode ; inode++) {
        u += element_p->nodes_p[inode]->u;
      }
      for (size_t inode = 0 ; inode < element_p->nnode ; inode++) {
        element_p->nodes_p[inode]->u = u/element_p->nnode;
      }
    }
  }

// ------------------------------------------------------------------- //
void Fem::_update_matrix() {
    for (auto& node : this->nodes) {
      node.mass = EV::Zero(node.dof);
      node.c    = EV::Zero(node.dof);
      node.dynamic_force = EV::Zero(node.dof);
    }

    for (auto& element : this-> elements) {
      element.set_xn();
      element.mk_local_update();

      size_t id = 0;
      for (size_t inode = 0 ; inode < element.nnode ; inode++) {
        for (size_t i = 0 ; i < this->dof ; i++) {
          element.nodes_p[inode]->mass[i] += element.M_diag[id];
          element.nodes_p[inode]->c[i] += element.C_diag[id];
          element.nodes_p[inode]->dynamic_force[i] += element.force[id];
          id++;
        }
      }
    }

    for (auto& node : this->nodes) {
      for (size_t i = 0 ; i < node.dof ; i++) {
        node.inv_mc[i] = 1.0 / (node.mass[i] + 0.5*dt*node.c[i]);
        node.mass_inv_mc[i] = node.mass[i] * node.inv_mc[i];
        node.c_inv_mc[i] = node.c[i] * node.inv_mc[i] * 0.5*dt;
        node.dtdt_inv_mc[i] = dt*dt*node.inv_mc[i];
      }
    }
  }
